src/web/components/StickyBottomBanner/bannerPicker.test.tsx

```diff
@@ -0,0 +1,195 @@
+import React from 'react';
+import { record } from '@root/src/web/browser/ophan/ophan';
+import { pickBanner, CanShowResult } from './bannerPicker';
+
+jest.mock('@root/src/web/browser/ophan/ophan', () => ({
+    record: jest.fn(),
+}));
+
+jest.useFakeTimers();
+
+// Wait for all unsettled promises to complete before finishing the test. Not
+// doing this results in a warning from Jest. Note that it's actually expected
+// that there may be outstanding promises when a test completes becuase the the
+// banner picker returns when the canShow of a banner resolves, even if there
+// are lower priority banners still pending.
+const flushPromises = () => new Promise(setImmediate);
+afterEach(async () => {
+    await flushPromises();
+});
+
+describe('pickBanner', () => {
+    it('resolves with the highest priority banner which can show', async () => {
+        const MockComponent = () => <div />;
+        const ChosenMockComponent = () => <div />;
+        const config = [
+            {
+                id: 'banner-1',
+                canShow: () => Promise.resolve({ result: false }),
+                show: () => MockComponent,
+                timeoutMillis: null,
+            },
+            {
+                id: 'banner-2',
+                canShow: () => Promise.resolve({ result: true }),
+                show: () => ChosenMockComponent,
+                timeoutMillis: null,
+            },
+            {
+                id: 'banner-3',
+                canShow: () => Promise.resolve({ result: true }),
+                show: () => MockComponent,
+                timeoutMillis: null,
+            },
+        ];
+
+        const got = await pickBanner(config);
+
+        expect(got()).toEqual(ChosenMockComponent);
```

Comment:
What does toEqual do with a component? Presumably checks by reference rather than by value!

Reply:
Hmm good question, yes I think that's correct. I guess the thing to note is that it's the component definition (e.g. `const ChosenMockComponent = () => { ... }` in the real world) that we're comparing here not a rendering of the component.